### PR TITLE
[Frontend_v2] Tab for participated challenges

### PR DESCRIPTION
**Pull request by Kajol-Kumari**
Fixes: #2698 [Add a tab for participated challenges for participants]

- [x] Added an API to fetch the participated challenges for a particular user.

- [x] Added a tab in the left sidebar to show the participated challenges for a participant.

GIF:
ezgif com-video-to-gif

### CODE REVIEW

**Comment by Kajol-Kumari** [comment]
@RishabhJain2018 @Sanji515 @lunayach can you please review this one?

File: frontend_v2/src/app/components/publiclists/challengelist/challengelist.component.ts [hunk 1]
```diff
@@ -334,4 +372,32 @@ export class ChallengelistComponent implements OnInit {
...
+      data => {
+        if (path.endsWith('future')) {
+          SELF.upcomingChallenges = data['results'];
+        } else if (path.endsWith('present')) {
+          SELF.ongoingChallenges = data['results'];
+        } else if (path.endsWith('past')) {
+          SELF.pastChallenges = data['results'];
+        }
+        SELF.filteredChallenges = SELF.upcomingChallenges.concat(SELF.ongoingChallenges, SELF.pastChallenges);
+        SELF.filteredOngoingChallenges = SELF.ongoingChallenges;
```
Comment (Sanji515) [inline frontend_v2/src/app/components/publiclists/challengelist/challengelist.component.ts:393]: As we are using same variable name in hosted challenges API also so do you think we should name different variable for participated challenges and the hosted challenges?

**Comment by Sanji515** [comment]
@Kajol-Kumari And yeah one more thing do we need to show the `upcoming participated challenges` tab as I don't think there will be such challenges because upcoming challenges are always inactive to participate.

cc @RishabhJain2018

File: frontend_v2/src/app/components/publiclists/challengelist/challengelist.component.ts (hunk 1 above)
Comment (Kajol-Kumari) [reply frontend_v2/src/app/components/publiclists/challengelist/challengelist.component.ts:393]: @Sanji515 we can name the different variables for participated and hosted challenges but i think it's fine this way also 'coz the variable will be assigned each time the required url is visited.

**Comment by RishabhJain2018** [comment]
> @Kajol-Kumari And yeah one more thing do we need to show the `upcoming participated challenges` tab as I don't think there will be such challenges because upcoming challenges are always inactive to participate.
> 
> cc @RishabhJain2018

Yes, I agree. It should be current participated challenge and past participated challenges. A participant can't register/participate in a challenge until the challenge is active.

**Comment by Kajol-Kumari** [comment]
Hey @Sanji515 can you please see if the changes looks good to you. If any other changes are required, please let me know.

File: apps/challenges/urls.py [hunk 2]
```diff
@@ -39,6 +39,11 @@
         views.get_all_challenges,
         name="get_all_challenges",
     ),
+    url(
+        r"^challenge/participated/(?P<challenge_time>[A-Za-z]+)$",
```
Comment (RishabhJain2018) [inline apps/challenges/urls.py:43]: `r"^challenge/participated/(?P<challenge_time>[A-Za-z]+)$",` -> `r"^challenges/participated/(?P<challenge_time>[A-Za-z]+)/$",`

Comment (Kajol-Kumari) [reply apps/challenges/urls.py]: @RishabhJain2018 I have made the suggested changes, please look at it.

File: frontend_v2/src/app/components/publiclists/challengelist/challengelist.component.ts [hunk 3]
```diff
@@ -212,6 +235,25 @@ export class ChallengelistComponent implements OnInit {
...
+   * Fetch participated challenges function.
+   * @param filter  selected filter
+   * @param callback  callback function
+   */
+  fetchMyParticipatedChallenges(filter = null, callback = null) {
+    if (!filter) {
+      const ALL_PATHS = Object.values(this.newApiPathMapping);
+      const ALL_KEYS = Object.keys(this.newApiPathMapping);
+      for (let i = 0; i < ALL_PATHS.length; i++) {
+        if (this[ALL_KEYS[i]]) {
```
Comment (lunayach) [inline frontend_v2/src/app/components/publiclists/challengelist/challengelist.component.ts:248]: Can you create variable SELF for "this"?

**Pull request merged**